archiver/service/archiver.go

```diff
@@ -0,0 +1,220 @@
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+const backfillErrorRetryInterval = 5 * time.Second
+
+var ErrAlreadyStopped = errors.New("already stopped")
+
+type BeaconClient interface {
+	client.BlobSidecarsProvider
+	client.BeaconBlockHeadersProvider
+}
+
+func NewService(l log.Logger, cfg flags.ArchiverConfig, dataStoreClient storage.DataStore, client BeaconClient, m metrics.Metricer) (*ArchiverService, error) {
+	return &ArchiverService{
+		log:             l,
+		cfg:             cfg,
+		dataStoreClient: dataStoreClient,
+		metrics:         m,
+		stopCh:          make(chan struct{}),
+		beaconClient:    client,
+	}, nil
+}
+
+type ArchiverService struct {
+	stopped         atomic.Bool
+	stopCh          chan struct{}
+	log             log.Logger
+	dataStoreClient storage.DataStore
+	beaconClient    BeaconClient
+	metricsServer   *httputil.HTTPServer
+	cfg             flags.ArchiverConfig
+	metrics         metrics.Metricer
+}
+
+func (a *ArchiverService) Start(ctx context.Context) error {
+	if a.cfg.MetricsConfig.Enabled {
+		a.log.Info("starting metrics server", "addr", a.cfg.MetricsConfig.ListenAddr, "port", a.cfg.MetricsConfig.ListenPort)
+		srv, err := opmetrics.StartServer(a.metrics.Registry(), a.cfg.MetricsConfig.ListenAddr, a.cfg.MetricsConfig.ListenPort)
+		if err != nil {
+			return err
+		}
+
+		a.log.Info("started metrics server", "addr", srv.Addr())
+		a.metricsServer = srv
+	}
+
+	currentBlob, _, err := a.persistBlobsForBlockToS3(ctx, "head")
+	if err != nil {
+		a.log.Error("failed to seed archiver with initial block", "err", err)
+		return err
+	}
+
+	go a.backfillBlobs(ctx, currentBlob)
+
+	return a.trackLatestBlocks(ctx)
+}
+
+func (a *ArchiverService) persistBlobsForBlockToS3(ctx context.Context, blockIdentifier string) (*v1.BeaconBlockHeader, bool, error) {
+	currentHeader, err := a.beaconClient.BeaconBlockHeader(ctx, &api.BeaconBlockHeaderOpts{
+		Block: blockIdentifier,
+	})
+
+	if err != nil {
+		a.log.Error("failed to fetch latest beacon block header", "err", err)
+		return nil, false, err
+	}
+
+	exists, err := a.dataStoreClient.Exists(ctx, common.Hash(currentHeader.Data.Root))
+	if err != nil {
+		a.log.Error("failed to check if blob exists", "err", err)
+		return nil, false, err
+	}
+
+	if exists {
+		a.log.Debug("blob already exists", "hash", currentHeader.Data.Root)
+		return currentHeader.Data, true, nil
+	}
+
+	blobSidecars, err := a.beaconClient.BlobSidecars(ctx, &api.BlobSidecarsOpts{
+		Block: currentHeader.Data.Root.String(),
+	})
+
+	if err != nil {
+		a.log.Error("failed to fetch blob sidecars", "err", err)
+		return nil, false, err
+	}
+
+	a.log.Debug("fetched blob sidecars", "count", len(blobSidecars.Data))
+
+	blobData := storage.BlobData{
+		Header: storage.Header{
+			BeaconBlockHash: common.Hash(currentHeader.Data.Root),
+		},
+		BlobSidecars: storage.BlobSidecars{Data: blobSidecars.Data},
+	}
+
+	err = a.dataStoreClient.Write(ctx, blobData)
```

Comment:
Can you document here (& in godoc) that you're assuming the L1 is fully trusted?  Or is there a verification process somewhere else that checks the blobs against the commitment and the EL blob hash matches the commitment?